detox/src/ipc/ipc.test.js

```diff
@@ -17,97 +20,164 @@ describe('IPC', () => {
+    describe('dispose()', () => {
+      it('should resolve if there are no connected clients', async () => {
+        await ipcServer.init();
+        await expect(ipcServer.dispose()).resolves.toBeUndefined();
+      });
 
-      await ipcClient1.dispose();
-      await ipcClient2.dispose();
+      it('should wait for connected clients to disconnect', async () => {
+        await ipcServer.init();
+        await ipcClient1.init();
 
-      setTimeout(() => {
-        throw new Error('dispose() should have resolved by now');
-      }, 3000);
+        const disposePromise = ipcServer.dispose();
+        const waitPromise = sleep(100).then(() => 'waited');
+        await expect(Promise.race([disposePromise, waitPromise])).resolves.toBe('waited');
 
-      await ipcServer.dispose();
+        await ipcClient1.dispose();
+        await expect(disposePromise).resolves.toBeUndefined();
+      });
+
+      it('should resolve if all clients have disconnected', async () => {
+        await ipcServer.init();
+
+        await ipcClient1.init();
+        await ipcClient2.init();
+
+        await ipcClient1.dispose();
+        await ipcClient2.dispose();
+
+        await ipcServer.dispose();
+      });
     });
   });
 
   describe('client', () => {
-    it('should return the correct session state', () => {
-      expect(ipcClient1.sessionState).toEqual(sessionState);
+    beforeEach(async () => {
+      await ipcServer.init();
+    });
+
+    it('should have a session state', () => {
+      expect(ipcClient1.sessionState).toEqual(new SessionState(sessionState));
     });
 
     it('should return the correct IPC server id', () => {
       expect(ipcClient1.serverId).toEqual(sessionState.detoxIPCServer);
     });
 
-    it('should throw if test results are reported before init', async () => {
-      const expected = 'IPC server bar has unexpectedly disconnected';
-      await expect(ipcClient1.reportTestResults([])).rejects.toThrow(expected);
-    });
+    describe('before init', () => {
+      it('should throw on attempt to report test results', async () => {
+        const expected = 'IPC server foo has unexpectedly disconnected';
+        await expect(ipcClient1.reportTestResults([])).rejects.toThrow(expected);
+      });
 
-    it('should throw if worker is registered before init', async () => {
-      const expected = 'IPC server bar has unexpectedly disconnected';
-      await expect(ipcClient1.registerWorker('foo')).rejects.toThrow(expected);
+      it('should throw on attempt to register the worker', async () => {
+        const expected = 'IPC server foo has unexpectedly disconnected';
+        await expect(ipcClient1.registerWorker('foo')).rejects.toThrow(expected);
+      });
     });
 
-    it('should register context upon init', async () => {
-      await ipcClient1.init();
-
-      let expected = ['foo'];
-      expect(ipcServer.sessionState.contexts).toEqual(expected);
-      expect(ipcClient1.sessionState.contexts).toEqual(expected);
```

Comment:
This was the test that should not have passed. Using the same instance of session state led to a false positive.

Reply:
:+1: